qvm-tools/qvm-sync-clock

```diff
@@ -40,19 +41,45 @@ def main():
             clockvm.name))
         sys.exit(0)
 
-    p = clockvm.run_service('qubes.GetDate')
-    untrusted_date_out = p.stdout.read(25).decode('ascii', errors='strict')
-    untrusted_date_out = untrusted_date_out.strip()
-
-    if not re.match(r'^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\+00:?00$', untrusted_date_out):
+    with clockvm.run_service('qubes.GetDate+nanoseconds') as p:
+        untrusted_date_out = p.stdout.read(36)
+    try:
+        untrusted_date_out.decode('ascii', 'strict')
```

Comment:
Please save result of this decode and operate on str further down. It makes sense to use `bytes` only for binary data (that can't be converted to `str`) or if you are not going to convert it to `str` at all. Otherwise `str` is preferred.

Reply:
Saving the result and passing it to the `date` command later on makes sense, but I am curious why the validation should be done on the `str` rather than on the `bytes`.  All of the needed operations work on both, and I suspect the `bytes` regex code is probably a bit simpler (less attack surface) than its `str` counterpart.  I could also just skip this, on the grounds that a non-ASCII date will fail to match any of the regexps and will thus rejected.

Reply:
> Saving the result and passing it to the `date` command later on makes sense, but I am curious why the validation should be done on the `str` rather than on the `bytes`.

Validation should be done on the thing you will actually use - so, since you'll need str, you should validate that str.
Theoretically you could do validation on bytes and convert only at the end, but there is no point in doing this (and surely forgetting some `b` at some point in the future).

Reply:
I decided to go with a different approach: take advantage of `subprocess` supporting `bytes` for command-line arguments.  The decode is now purely an input validation step; its output is not used.